### PR TITLE
refactor transformer decoder and revamp the left padding attention mask

### PR DESCRIPTION
**Pull request by vince62s**
The initial goal was to gather transformer_lm and transformer decoder type in a single class
 - remove decoder_type = "transformer_lm"
 - we keep it at "architecture" level = ["transformer", "transformer_lm"]

After carefull review I realized that left padding was not handled 100% properly.
After this PR:

For training:
we rely on pytorch `scaled_dot_product_attention` in `memory_efficient` mode.
we may need to investigate the `flash_varlen_func()` (which seems slower for now). see https://github.com/Dao-AILab/flash-attention/issues/1125
==> `self_attn_backend`: has no impact at training at the moment.

For inference:
For EncoderDecoder models we rely also on pytorch `scaled_dot_product_attention` in `memory_efficient` mode since it is much faster than `flash_attn_with_kvcache()` due to short sequences / large batches.
For LLMs (Decoder only models) we rely on `flash_attn_with_kvcache()` which much faster.

When we use SDPA we always buid ourselves the attention_mask and never rely on is_causal=True
The main reason is that we need to handle padding properly and not attend padded tokens.
With `flash_attn_with_kvcache()` we added a "recent" functionality `cache_leftpad` we requires flash-attn > 2.6.1

Results are much more consistent between the two options now.

==> `self_attn_backend`: if "flash" then `flash_attn_with_kvcache()` will be used, if "pytorch" then sdpa in memory_efficient mode will be used.

### CODE REVIEW

**Review by francoishernandez — COMMENTED** [review]
Nicely done! This simplification had been overdue for quite some time.
(Did not test changes on my end yet, but I trust the existing tests should cover most of the changes here.)

Minor comments to address.
Also, for reference, what is the rationale behind embedding the `left_pad` flag within the batch?

File: eole/bin/run/predict.py [hunk 1]
```diff
@@ -36,6 +37,10 @@ def predict(config):
 class Predict(RunBin):
     config_class = PredictConfig
     require_config = False
+    torch.backends.cuda.enable_mem_efficient_sdp(True)
+    torch.backends.cuda.enable_flash_sdp(False)
+    torch.backends.cuda.enable_math_sdp(False)
+    torch.backends.cuda.enable_cudnn_sdp(False)
```
[the remark this answers is not in the document]
Comment (vince62s) [reply eole/bin/run/predict.py:43]: I guess you're correct, I thought it was better to make things clear from the very start.

**Comment by vince62s** [comment]
> Also, for reference, what is the rationale behind embedding the left_pad flag within the batch?

When using left_padding we need to mask the pad tokens at EACH time step when decoding.
So knowing your batch is in left_pad mode triggers that at step 0 you store the pad token mask (in the prompt) and you re-use it at each time step decoding.
